### ISSUE TITLE
Smarter value-based equality is necessary for supporting OO code

### ISSUE THREAD

**Issue by asaf-romano** [ramda/ramda#1383]
The current value-based equality check done by Ramda (across the board) interacts badly with one of those "best practices" of OO in JS, which is to make private properties non-enumerable, or, in es6 environment, be set as symbols.

non-enumerable property problem demo: http://bit.ly/1UR0Bhc
symbol property problem demo: http://bit.ly/1Oqc6wd

I really do like value-based equality, and I also acknowledge that supporting OOP code in general is not the main focus of Ramda. I also acknowledge that there are probably some cases involving js closures (and/or WeakMaps tricks) which are simply unfixable, and also that the common implementations of deepEqual are affected by the very same issue.

However, I believe this issue is going to surprise a lot of people, esp. when dealing with 3rd party objects - e.g. running uniq on  an array of C objects (see demos) is always going to returns just the first element. If you're dealing with your own library and know ramda quite well, the option of dispatching is a fine workaround, but when dealing with third party libraries, more often than not you're out of luck. Given ES6 classes syntax (implement in current chromium nightlies and soon in Fx) along with upcoming ES7 decorators, symbols/non-enumerable properties as "privatizing" mechanism is going to spread out a lot, rendering this issue quite severe. Given all that, my hunch would be that the deep equality check should also take non-enumerable properties and symbol properties into account (I'd not do that for isEmpty though...). Indeed, there's a trade-of here, because it'd break the current equality check in some cases (``moment.js`` for example... I don't remember the details).

**Comment by davidchambers** [comment]
Something I've been considering for a while is checking `a.constructor === b.constructor` (or `a.constructor.name === b.constructor.name`) when comparing `a` and `b`. A `Foo` instance with no enumerable properties is not equivalent to a `Bar` instance with no enumerable properties.

**Comment by asaf-romano** [comment]
Indeed, but note that's a different use case. Here the constructor is the same.

BTW, Function.name is still missing from IE, so it's better to use getPrototypeOf for that.

**Comment by asaf-romano** [comment]
BTW, ``toString`` is also affected by this issue. I'd not care about that if memoize wasn't relying on it.

**Comment by CrossEye** [comment]
> The current value-based equality check done by Ramda (across the board) interacts badly with one of those "best practices" of OO in JS, which is to make private properties non-enumerable, or, in es6 environment, be set as symbols. [...] I really do like value-based equality, and I also acknowledge that supporting OOP code in general is not the main focus of Ramda. I also acknowledge that there are probably some cases involving js closures (and/or WeakMaps tricks) which are simply unfixable...

It's the latter that worries me.  OO is done in JS in so many ways.  Perhaps ES6 will eventually bring about some consistency, but I'm not particularly hopeful.  Here's a simple case that I don't think we could ever solve:

```js
function Rectangle(w, h, s) {
  var scale = arguments.length > 2 ? s : 1;
  this.width = w;
  this.height = h;
  this.area = function() {return this.width * this.height * scale;}
}

var r1 = new Rectangle(5, 3, 1.0);
var r2 = new Rectangle(5, 3, 1.1);
R.equals(r1, r2);  // true
R.equals(r1.area(), r2.area()); // false
```

If we can't solve some of the fundamental forms of OO, I'm suspicious of the project of trying to capture other ones.  I think we'll end up ever chasing an elusive, and finally impossible, goal.

I like the dispatching approach, which allows the user to determine how `equals` applies to her class of objects.

So, while I'm curious to hear further discussion, my first thought is negative.

**Comment by asaf-romano** [comment]
I'm not suggesting to solve all those cases. As I said, I'm worried about "best practices" being broken. non-enumerable property is the closest thing to a private member property, and many good libraries use it this way. Symbols are first-class citizens in ES6 that Ramda simply ignores (even though there's a way to get them). With babel.js and typescript rapidly popularizing ES6 classes and decorators, very self-descriptive code that doesn't look hackish like the above at all is now broken. I think that a complete solution here involves updating the documentation of equals to make its limitations clear and to promote the dispatching workaround there (And also note that you don't really want to dispatch to ``equals`` without also overriding ``toString`` because then funny things happen).

I think it's possible to draw a border line between the cases we want to solve and the cases we don't. Getters vs. methods is an easy one, imo: your example is basically a non-getter getter, which, without AI, ramda cannot distinguish from normal methods. However, t should always be fine for ramda to invoke getters, so you'd have caught that if ``area`` was defined as a getter. If someone breaks the getter "contract" (``foo.initialized;  // initialize the object``) then there's nothing we can do.

**Comment by CrossEye** [comment]
Just to make it clear, I'm not actually opposed to updating equals to take into account non-enumerable properties and symbols.  While not 100% certain, my first impression is that it would be a good idea.

What I'm worried about it trying to chase down the OO technique of the month.  What you're calling "best practices", a second developer might find horrible, and might never even occur to a third one.  Yes, the new syntax from ES6 may in time help settle this debate, but at the moment it's far from clear that it's true.  My example was not about the `area` function; it was about the actual piece of data that distinguishes the two values, the `scale` closure property.  This is not at all an unusual way to deal with concerns for privacy in OO code, and nothing we do can overcome that.  But the developer of the OO system might well be able to do so, which is why my answer to OO-specific questions for such functions is delegation.

But to the main point, expanding `equals` to include non-enumerable properties and symbols, my vote is a tentative yes.

**Comment by asaf-romano** [comment]
OK, I updated the summary to reflect the part we can agree on. I think this is good enough.

But regarding the area function, I think I made my point poorly. Let's view it from a different POV. Given the follow class:
```js

class Foo {
   constructor()
   {
       this.publicFieldFromCtor = null;
   }
   get getterA() {
     ...
   }

   get getterB()
   {

   }

   publicES7Field = 3;

   functionThatIsActuallyAGetter()
   {
      ....
   }
}
```

JS internals and prototype-OO quirks aside, I think that it's very intuitive for a default equals implementation, esp. one that is done "from the outside" (the way ramda does) to compare getterA, getterB and marginally less the the two public fields. Intuitive as it is, it's very hard to implement because the two getters are not set as properties of instances of this class, while the "fields" are. The getters are always at least one level up in the prototype chain. Even worse, in ES6 classes the are marked as non-enumerable. So while the "contract" parts of the class (i.e. the getters) is, at least to me, the most obvious candidate for an equality check, they are the hardest to get right without getting a lot of noise and hurting performance quite a bit (by walking the prototype chain). ``functionThatIsActuallyAGetter`` (the analogy to area) is an interesting problem, but there's nothing we can do about it, so let's leave it out of this discussion.

So what Ramda assumes (with or without the suggested fix), in par with may other deep equality implementations, in its ``equals`` implementation, is that if it covers an object's own properties, it'll, most of the time also cover what matters. Given all considerations, I think this is approach is fine, but it does mean you've to take private-ish fields into account. Otherwise you're very likely to miss the public getters much more often.

**Comment by CrossEye** [comment]
I do think that we'll have to think very carefully before deciding to tread here.  I worry about the number of crazy edge cases there are to consider.

**Comment by asaf-romano** [comment]
BTW, I'm not sure what R.clone does exactly, but it seems that it gets what I described in my previous comment quite nicely (symbols aside)... it flattens the getters from the prototype chain.

The equality check that returned true (wrongly) returns false for clones
http://bit.ly/1iM224w

Interesting and terrible at the same time :)

**Comment by paldepind** [comment]
I would like to point out that many Ramda functions use `equals`. If we make `equals` slower many other functions will be slower as well.

**Comment by asaf-romano** [comment]
Indeed, but see #1386. I think most, if not all, all of the not-O(n) cases could be addressed this way.
